ext_mongodb.php

```diff
@@ -378,65 +378,229 @@ public function __debugInfo()
+			return true;
+		}
 
-			if (array_key_exists('modifiers', $options)) {
-				Utils::mustBeArrayOrObject('modifiers', $options['modifiers']);
-				foreach ($options['modifiers'] as $key => $value) {
-					$this->query['query'][$key] = $value;
-				}
+		return false;
+	}
+
+	/* Helpers for setting / validating complex options */
+	private function _queryInitHint( array $options, array $modifiers )
+	{
+		if ( array_key_exists( 'hint', $options ) )
+		{
+			if ( is_string( $options['hint'] ) )
+			{
+				$this->_queryOptString( 'hint', $options, 'hint' );
+			}
+			else if ( is_object( $options['hint'] ) || is_array( $options['hint'] ) )
+			{
+				$this->_queryOptDocument( 'hint', $options, 'hint' );
+			}
+			else
+			{
+				throw new \MongoDB\Driver\Exception\InvalidArgumentException(
+					sprintf(
+						'Expected "hint" option to be string, array, or object, %s given',
+						gettype( $options['hint'] )
+					)
+				);
+			}
+		}
+		else if ( array_key_exists( '$hint', $modifiers ) )
+		{
+			if ( is_string( $modifiers['$hint'] ) )
+			{
+				$this->_queryOptString( 'hint', $modifiers, '$hint' );
 			}
+			else if ( is_object( $modifiers['$hint'] ) || is_array( $modifiers['$hint'] ) )
+			{
+				$this->_queryOptDocument( 'hint', $modifiers, '$hint' );
+			}
+			else
+			{
+				throw new \MongoDB\Driver\Exception\InvalidArgumentException(
+					sprintf(
+						'Expected "$hint" modifier to be string, array, or object, %s given',
+						gettype( $modifiers['$hint'] )
+					)
+				);
+			}
+		}
+	}
 
-			if (array_key_exists('projection', $options)) {
-				Utils::mustBeArrayOrObject('projection', $options['projection']);
-				$this->query['fields'] = (array) $options['projection'];
+	private function _queryInitLimitAndSingleBatch( array $options )
+	{
+		if ( array_key_exists( 'limit', $options ) && $options['limit'] < 0 )
+		{
+			$this->opts['limit'] = 0 - $options['limit'];
+
+			if ( array_key_exists( 'singleBatch', $options ) && !$options['singleBatch'] )
+			{
+				throw new \MongoDB\Driver\Exception\InvalidArgumentException( 'Negative "limit" option conflicts with false "singleBatch" option' );
 			}
+			else
+			{
+				$this->opts['singleBatch'] = true;
+			}
+		}
+		else
+		{
+			$this->_queryOptInt64( 'limit', $options, 'limit' );
+			$this->_queryOptBool( 'singleBatch', $options, 'singleBatch' );
+		}
+	}
 
-			if (array_key_exists('sort', $options)) {
-				Utils::mustBeArrayOrObject('sort', $options['sort']);
-				$this->query['query']['$orderby'] = (object) $options['sort'];
+	private function _queryInitReadConcern( array $options )
+	{
+		if ( array_key_exists( 'readConcern', $options ) )
+		{
+			if ( !is_object( $options['readConcern'] ) || ! $options['readConcern'] instanceof \MongoDB\Driver\ReadConcern )
+			{
+				throw new \MongoDB\Driver\Exception\InvalidArgumentException(
+					sprintf(
+						'Expected "readConcern" option to be %s, %s given',
+						'MongoDB\Driver\ReadConcern',
```

Comment:
Was `"MongoDB\Driver\ReadConcern"` an argument just to help with wrapping? I don't see a reason to separate it otherwise.